### PR TITLE
Epic: Mixin Elements Core

### PR DESCRIPTION
**Pull request by ala-n**
Here is a draft implementation for #671 

Feel free to post your thoughts/suggestions here and in PM.

### CODE REVIEW

**Review by dshovchko — APPROVED** [review]

File: src/modules/esl-base-element/core/esl-mixin-registry.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+// Private key to store mixin instances
+const STORE = '__mixin';
+
+// Singleton for registry
+let global: ESLMixinRegistry;
+
+/** Registry to store and initialize {@link ESLMixinElement} instances */
+export class ESLMixinRegistry {
+  /** Map that sores available mixins under their identifier (attribute) */
```
Comment (NastaLeo) [inline src/modules/esl-base-element/core/esl-mixin-registry.ts:11]: Typo: stores

**Review by NastaLeo — APPROVED** [review]

File: src/modules/esl-base-element/core/esl-mixin-registry.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    return instance;
+  }
+  /** Destroys passed mixin on the element */
+  private static destroy(el: HTMLElement, mixin: string): void {
+    const store = (el as any)[STORE] as Record<string, ESLMixinElement>;
+    if (!store) return;
+    const instance = store[mixin];
+    if (!instance) return;
+    instance.disconnectedCallback();
+    delete store[mixin];
```
Comment (abarmina) [inline src/modules/esl-base-element/core/esl-mixin-registry.ts:139]: why don't we use the variable `instance` here?

**Review by abarmina — APPROVED** [review]

File: src/modules/esl-base-element/core/esl-mixin-registry.ts (hunk 2 above)
Comment (abarmina) [reply src/modules/esl-base-element/core/esl-mixin-registry.ts:139]: I got it)

**Review by yadamskaya — APPROVED** [review]

File: src/modules/esl-base-element/README.md [hunk 3]
```diff
@@ -1,28 +1,57 @@
...
 
-Version: *1.0.0*
+Version: *1.1.0*
 
 Authors: *Alexey Stsefanovich (ala'n)*
 
 <a name="intro"></a>
 
-Provides the ESLBaseElement - a base class for custom element declaration, and a set of common decorators.
+Provides the core to create custom elements and ts decorators to simplify components creation.
```
Comment (NastaLeo) [inline src/modules/esl-base-element/README.md:9]: Provides the core to create custom elements and ts decorators to simplify components developing.

File: src/modules/esl-base-element/README.md [hunk 4]
```diff
@@ -1,28 +1,57 @@
...
+### Base Element API
+- `connected` - readonly marker, true if element connected and base `connectedCallback` executed
+
+### Mixin Element API (beta)
+- `$host` - readonly mixin target DOM element
+
+### Element decorators:
+Works for both `ESLBaseElement` and `ESLMixinElement`.
+
+ - `@attr` - to map string type property to HTML attribute.
```
Comment (NastaLeo) [inline src/modules/esl-base-element/README.md:46]: As I remember we don't use a dot in such type of phrase (2 following phrases also)

File: src/modules/esl-base-element/README.md [hunk 5]
```diff
@@ -54,3 +83,33 @@ MyCustomComponent.register();
 // Or register custom tag with passed tag name
 MyCustomComponent.register('my-tag');
 ```
+
+### Mixin Example (beta)
+
+```ts
+import {ESLMixinElement, attr, boolAttr, jsonAttr} from '@exadel/esl';
+
+class MyMixiinComponent extends ESLMixinElement {
```
Comment (NastaLeo) [inline src/modules/esl-base-element/README.md:92]: class MyMixinComponent extends ESLMixinElement

**Review by NastaLeo — APPROVED** [review]

**Review by abarmina — APPROVED** [review]

**Review by julia-murashko — APPROVED** [review]

**Review by abarmina — APPROVED** [review]

**Review by nattallius — APPROVED** [review]

**Pull request merged**